src/domain/model/user/Address.test.ts

```diff
@@ -0,0 +1,85 @@
+import { Address } from './Address';
+import { Country } from './Country';
+
+describe('Address', () => {
+  it('creates a valid address', () => {
+    expect(() => {
+      new Address(
+        'address1',
+        'address2',
+        'city',
+        'region',
+        'postcode',
+        new Country('GB')
+      );
+    }).not.toThrow();
+
+    expect(() => {
+      new Address(
+        'address1',
+        undefined,
+        'city',
+        undefined,
+        'postcode',
+        new Country('GB')
+      );
+    }).not.toThrow();
+  });
+
+  it('does not accept empty strings', () => {
+    expect(() => {
+      new Address(
+        '',
+        'address2',
+        'city',
+        'region',
+        'postcode',
+        new Country('GB')
+      );
+    }).toThrow();
+
+    expect(() => {
+      new Address(
+        'address1',
+        '',
+        'city',
+        'region',
+        'postcode',
+        new Country('GB')
+      );
+    }).toThrow();
```

Comment:
Why should this throw?  From a client perspective it's much easier to send the empty strings than not, because the form will naturally turn an empty text input into an empty string.  At some point, wither the client or  application will need to strip those out, right?

Reply:
Yeah I see the point. Ideally we should have a canonical form of undefined regions in the database. (It shouldn't be that some are undefined and some are empty strings). With that in mind the domain object shouldn't let things be either or. 

But it's a good comment. We probably need to handle that on the application layer..